src/mame/namco/namcos10.cpp

```diff
@@ -2595,9 +2618,143 @@ void namcos10_memp3_state::namcos10_memp3_map(address_map &map)
+}
+
+uint16_t namcos10_memp3_state::dram_pld_ver_r()
+{
+	// DRAM-C PLD version, only seems to be read when sub CPU is in debug mode
+	return 1;
+}
+
+uint16_t namcos10_memp3_state::nand_pld_ver_r()
+{
+	// NAND-C PLD version, only seems to be read when sub CPU is in debug mode
+	return 1;
+}
+
+uint16_t namcos10_memp3_state::mp3_stream_status_r()
+{
+	// The debug messages when debug mode are enabled checks this register
+	// and if it's 0 then it prints "MUTE:ON" and if it's 1 it prints "MUTE:OFF"
+	return m_mp3_stream_available;
+}
+
+void namcos10_memp3_state::mp3_stream_status_w(uint16_t data)
+{
+	// This gets set to 0 and then 1 every time a new MP3 stream is started
+	bool is_available = BIT(data, 0);
+
+	if (m_mp3_stream_available && !is_available)
+		m_lc82310->reset_playback();
+
+	m_mp3_stream_available = is_available;
+}
+
+void namcos10_memp3_state::mp3_data_w(uint16_t data)
+{
+	m_lc82310->dimpg_w(data & 0xff);
+	m_mp3_received_byte = true;
 }
 
+void namcos10_memp3_state::subcomm_ack_w(uint16_t data)
+{
+	m_subcomm_busy = false;
+	m_memp3_mcu->set_input_line(TLCS900_INT5, CLEAR_LINE);
+}
+
+uint16_t namcos10_memp3_state::mp3_unk_r()
+{
+	// Seems to be related to timer 3, something to do with the "L" LEDs maybe?
+	// When the timer 3 value has ticked 10 times, set to 1 if value is 0 else shift value by 1
+	// mp3_unk_w(mp3_unk_r() ? mp3_unk_r() << 1 : 1)
+	return m_mp3_unk_val;
+}
+
+void namcos10_memp3_state::mp3_unk_w(uint16_t data)
+{
+	m_mp3_unk_val = data;
+}
+
+uint16_t namcos10_memp3_state::mp3_unk2_r()
+{
+	// The following are how the sub CPU program handles the bits.
+	// It's unclear what writing to these registers actually does outside of the sub CPU.
+	// Only available in g13jnr and nicetsuk, maybe some kind of debug controls?
+	// bit 0 = Start playback
+	// bit 1 = Stop playback
+	// bit 2 = Fade out
+	// bit 3 = Fade in
+	// bit 4 = Decrease volume by 1
+	// bit 5 = Increase volume by 1
+	// bit 6 = Mute/Unmute (uses 0x100032)
+	// bit 7 = Print state from MP3 decoder + state byte from sub CPU to the sub CPU's serial output
+	return m_mp3_unk2_val;
+}
+
+void namcos10_memp3_state::mp3_unk2_w(uint16_t data)
+{
+	m_mp3_unk2_val = data;
+}
+
+uint16_t namcos10_memp3_state::mp3_unk3_r()
+{
+	// When the timer 3 value has ticked 10 times: mp3_unk3_w(3 - mp3_unk3_r())
+	return m_mp3_unk3_val;
+}
+
+void namcos10_memp3_state::mp3_unk3_w(uint16_t data)
+{
+	m_mp3_unk3_val = data;
+}
+
+template <int Port>
+void namcos10_memp3_state::port_write(offs_t offset, uint8_t data)
+{
+	if (Port == 7) {
+		m_lc82310->zcsctl_w(BIT(data, 3));
+		m_lc82310->dictl_w(BIT(data, 1));
+		m_lc82310->ckctl_w(BIT(data, 2));
+		m_mp3_port7_data = data;
+	} else if (Port == 10) {
+		// bit 2 is toggled off/on after MP3 byte is written
+		// bit 3 is set to 0 before writing data and then set to 1 after writing data
+		m_mp3_porta_data = data;
+	}
+}
+
+template <int Port>
+uint8_t namcos10_memp3_state::port_read(offs_t offset)
+{
+	auto r = 0;
```

Comment:
Please use an explicit type for this.  Using `auto` like this makes it `int` but you should be using some unsigned integer type.